[flang]Add new intrinsic function backtrace and complete the TODO of abort

dty2 opened the pull request:
Hey guys, I found that Flang's built-in ABORT function is incomplete when I was using it. Compared with gfortran's ABORT (which can both abort and print out a backtrace), flang's ABORT implementation lacks the function of printing out a backtrace. This feature is essential for debugging and understanding the call stack at the failure point.

To solve this problem, I completed the "// TODO:" of the abort function, and then implemented an additional built-in function BACKTRACE for flang. After a brief reading of the relevant source code, I used backtrace and backtrace_symbols in "execinfo.h" to quickly implement this. But since I used the above two functions directly, my implementation is slightly different from gfortran's implementation (in the output, the function call stack before main is additionally output, and the function line number is missing). In addition, since I used the above two functions, I did not need to add -g to embed debug information into the ELF file, but needed -rdynamic to ensure that the symbols are added to the dynamic symbol table (so that the function name will be printed out).

Here is a comparison of the output between gfortran 's backtrace and my implementation:
gfortran's implemention output:
```
#0  0x557eb71f4184 in testfun2_
        at /home/hunter/plct/fortran/test.f90:5
#1  0x557eb71f4165 in testfun1_
        at /home/hunter/plct/fortran/test.f90:13
#2  0x557eb71f4192 in test_backtrace
        at /home/hunter/plct/fortran/test.f90:17
#3  0x557eb71f41ce in main
        at /home/hunter/plct/fortran/test.f90:18
```
my impelmention output:
```
Backtrace:
#0 ./test(_FortranABacktrace+0x32) [0x574f07efcf92]
#1 ./test(testfun2_+0x14) [0x574f07efc7b4]
#2 ./test(testfun1_+0xd) [0x574f07efc7cd]
#3 ./test(_QQmain+0x9) [0x574f07efc7e9]
#4 ./test(main+0x12) [0x574f07efc802]
#5 /usr/lib/libc.so.6(+0x25e08) [0x76954694fe08]
#6 /usr/lib/libc.so.6(__libc_start_main+0x8c) [0x76954694fecc]
#7 ./test(_start+0x25) [0x574f07efc6c5]
```
test program is:
```
function testfun2() result(err)
  implicit none
  integer :: err
  err = 1
  call backtrace
end function testfun2

subroutine testfun1()
  implicit none
  integer :: err
  integer :: testfun2

  err = testfun2()
end subroutine testfun1

program test_backtrace
  call testfun1()
end program test_backtrace
```
I am well aware of the importance of line numbers, so I am now working on implementing line numbers (by parsing DWARF information) and supporting cross-platform (Windows) support.

dty2 committed:
[flang]Add new intrinsic function backtrace and complete the TODO of abort

tblah commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +153,29 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
 
+static void PrintBacktrace() {
+  // TODO: Need to parse DWARF information to print function line numbers
+  const int MAX_CALL_STACK = 999;
```
In the runtime we use braced initialization. I also changed this to `constexpr`.
```suggestion
  constexpr int MAX_CALL_STACK{999};
```

tblah commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +153,29 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
+  // TODO: Need to parse DWARF information to print function line numbers
+  const int MAX_CALL_STACK = 999;
+  void *buffer[MAX_CALL_STACK];
+  int nptrs = backtrace(buffer, MAX_CALL_STACK);
```
```suggestion
  int nptrs{backtrace(buffer, MAX_CALL_STACK)};
```

tblah commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +153,29 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
+  const int MAX_CALL_STACK = 999;
+  void *buffer[MAX_CALL_STACK];
+  int nptrs = backtrace(buffer, MAX_CALL_STACK);
+  char **symbols = backtrace_symbols(buffer, nptrs);
```
```suggestion
  char **symbols{backtrace_symbols(buffer, nptrs)};
```

tblah commented on flang/runtime/stop.cpp:
```diff
@@ -15,6 +15,7 @@
 #include <cfenv>
 #include <cstdio>
 #include <cstdlib>
+#include <execinfo.h>
```
So far as I know this not a standard header - it is a GNU extension.

Flang supports other systems such as AIX, MacOS, Windows, and even other C libraries on gnu+linux systems (e.g. llvm-libc).

Please could you put this behind a preprocessor check so that compilation is not broken on those other systems.

dty2 replied:
You're right, I'll change this asap.

dty2 replied:
Hi, thank you very much for your suggestion. I have modified the code. You can check the latest commit. If possible, please give LGTM. Thank you.

kparzysz replied:
We have a config file with the results of the checks made by cmake: `<build-dir>/include/llvm/Config/config.h`.  One of these checks is for the `backtrace` function, and the header that declares it.
Please add
```
#include "llvm/Config/config.h"
```
then replace the `#ifdef __linux__` part with
```
#ifdef HAVE_BACKTRACE
#include BACKTRACE_HEADER
#endif
```

tblah reviewed (commented):
Thank you for implementing this.

kparzysz reviewed (commented):
This is all good, could you restore the branch and reopen the PR?

kparzysz commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +153,29 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
 }
 
+static void PrintBacktrace() {
+  // TODO: Need to parse DWARF information to print function line numbers
```
No worries.

kparzysz replied:
I guess whether we want to link Support into flang-rt may need a broader discussion.  Until we sort it out, we can keep your function.

dty2 replied:
I tried to reopen this PR, but I had deleted this branch before, so I plan to submit a PR again later, and I will also change some of the content that the boss above helped me modify. Thank you very much for your reply

dty2 replied:
I have reverted the branch and reopened this PR, thank you very much for your reply

dty2 committed:
commit some suggestions

dty2 committed:
Merge branch 'main' into dty2

dty2 commented:
@kparzysz Could you please review the latest changes? I've addressed all the feedback from the previous review. Thank you!

dty2 commented:
@kparzysz Could you please review the latest changes? I've addressed all the feedback from the previous review. Thank you! And I don't have merge permission. I hope you can merge the code if there is no problem after checking. Thank you very much.

kparzysz commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +156,39 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
 }
 
+static void PrintBacktrace() {
+#ifdef __linux__
```
Please replace it with `#ifdef HAVE_BACKTRACE`.

kparzysz commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +156,39 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
+
+  if (symbols == nullptr) {
+    Fortran::runtime::Terminator{}.Crash("no symbols");
+    std::exit(EXIT_FAILURE);
```
We shouldn't exit here, just not print any symbols and return from the function.  You can just do
```
if (char **symbols{...}) {
  for (...) {
    print symbol[i]
  }
  free(symbols);
}
```

kparzysz commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +156,39 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
+
+#else
+
+  // TODO: Windows platform implemention
```
Typo: implemention -> implementation

kparzysz commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +156,39 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
+}
+
 [[noreturn]] void RTNAME(Abort)() {
-  // TODO: Add backtrace call, unless with `-fno-backtrace`.
```
We still need a check for the `-fno-backtrace` option.  That can be done in another PR.  Will you be able to do that?

dty2 replied:
Yes, I will try to implement it.

dty2 committed:
[flang]Import the header file config.h in flang/runtime/stop.cpp and delete std::exit in PrintBacktrace

dty2 committed:
Merge branch 'dty2' of github.com:dty2/llvm-project into dty2

dty2 committed:
Merge branch 'main' into dty2

dty2 commented:
@kparzysz  Thank you for your patience in guiding my code. Please review the latest changes again to see if they meet the merge criteria. I will try my best to implement the -fno-backtrace option. If the implementation is successful, I will submit another PR. And I will continue to do my part for the llvm and flang communities.

kparzysz commented:
Looks good to me.

dty2 commented:
Oh, thanks, Can you merge this code? I don't have merge permissions

clementval commented on flang/include/flang/Runtime/stop.h:
```diff
@@ -29,6 +29,7 @@ NORETURN void RTNAME(ProgramEndStatement)(NO_ARGUMENTS);
 // Extensions
 NORETURN void RTNAME(Exit)(int status DEFAULT_VALUE(EXIT_SUCCESS));
 NORETURN void RTNAME(Abort)(NO_ARGUMENTS);
+void RTNAME(Backtrace)(NO_ARGUMENTS);
```
```suggestion
NORETURN  void RTNAME(Backtrace)(NO_ARGUMENTS);
```

dty2 replied:
Thank you for your suggestion. But I don't quite understand it, because as far as I know, this macro is used when the function will cause the program to terminate or will not return to the caller. The Backtrace built-in function itself does not cause the program to terminate, and this built-in function will return to the caller function to continue execution. Therefore, I don't think it is necessary to add this macro. Of course, if there is another reason, please let me know, and I will accept your suggestion.

clementval replied:
Ok if the backtrace returns then you don't need it.

tblah reviewed (approved):
LGTM. If nobody objects I can merge this tomorrow.

Thanks for all of the quick responses to feedback.

tarunprabhu commented on flang/runtime/stop.cpp:
```diff
@@ -152,11 +157,34 @@ void RTNAME(PauseStatementText)(const char *code, std::size_t length) {
+
+#else
+
+  // TODO: Windows platform implementation
```
Is this comment still accurate? Should it instead say "Handle the case when a backtrace is not available", or something along those lines? 

Does it makes sense to print a message saying that a backtrace was not available instead? It probably does not matter in the case of `abort`, but if the non-standard `backtrace` extension is used, it may be confusing if, it doesn't print anything on some platform".

dty2 replied:
I think you are right, doing nothing does lead to confusion, I will add this, thanks for the suggestion

dty2 replied:
I have modified this point. Thank you again for your suggestion. It is really necessary.

dty2 committed:
[flang]Handle the case when a backtrace is not available

dty2 committed:
Merge branch 'dty2' of github.com:dty2/llvm-project into dty2

dty2 committed:
Merge branch 'main' into dty2

kparzysz reviewed (approved):
LGTM

tarunprabhu reviewed (approved):
It looks [like you need to run clang-format](https://github.com/llvm/llvm-project/pull/117603#issuecomment-2503601394). Please make sure that the pre-merge checks pass before merging. 

Other than that, LGTM. Thanks for seeing this through.

dty2 committed:
[flang]update format

dty2 committed:
Merge branch 'dty2' of github.com:dty2/llvm-project into dty2

dty2 committed:
Merge branch 'main' into dty2

dty2 committed:
Merge branch 'main' into dty2

tblah commented:
I reverted due to the failed buildbot. The buildbot failed because the LLVM configuration header was missing.

The buildbot had a separate step for building the runtime so I wonder if it built only that and nothing in `llvm/`. So the problem could be a missing cmake dependency from the runtime to whatever in `llvm/` generates that header. 

It looks like there is already a `$BUILD_DIR/flang/runtime/config.h` generated from `$SOURCE_DIR/flang/runtime/config.h.cmake`. I think you will need to add the right line to that cmake file so that this variable is set, then include the flang runtime version of `config.h`. You can find the line you need in `llvm/include/llvm/Config/config.h.cmake`.

Thanks for all of your work so far :) don't hesitate to reach out if you have any questions.

jeanPerier commented on flang/lib/Evaluate/intrinsics.cpp:
```diff
@@ -1336,6 +1336,7 @@ static const IntrinsicInterface intrinsicSubroutine[]{
             {"stat", AnyInt, Rank::scalar, Optionality::optional,
                 common::Intent::Out}},
         {}, Rank::elemental, IntrinsicClass::atomicSubroutine},
+    {"backtrace", {}, {}, Rank::elemental, IntrinsicClass::pureSubroutine},
```
I am late in the review, but since this has been reverted, I do not think "elemental" is correct here, and I would also debate the pure aspect since this is doing IO on the standard output which is forbidden for pure subroutine.

In fact, I am not quite sure that the front-end needs to know about backtrace since there are type resolution or argument requiring explicit interface.

It may just be simpler to deal with it like a user procedure for which the runtime provides an implementation (if the user has not). See [FDATE implementation](https://github.com/llvm/llvm-project/pull/71222/files) for instance.

Please also document the extension support in docs/Intrinsics.md. 

Thanks for adding this, it is a great feature.

dty2 replied:
Oh! I understand what you said. I think you are right. But in this case, it seems to me that the built-in functions such as Abort and Exit need to be rewritten, because they can be regarded as user processes like backtrace. (If I understand correctly) I plan to complete backtrace in the original way first, and rewrite the built-in functions such as Abort and exit in the next new PR.

dty2 replied:
I temporarily changed pure to impure to match element (I saw the original Abort implementation was the same). Since the previous PR was problematic and had been rejected, I submitted a new PR after making modifications. You can follow this PR to follow the latest changes.

dty2 replied:
I temporarily changed pure to impure to match element (I saw the original Abort implementation was the same). Since the previous PR was problematic and had been rejected, I submitted a new PR after making modifications. You can follow this PR to follow the latest changes.#118179

klausler replied:
Agree.
